adapters/repos/db/vector/hnsw/compress.go

```diff
@@ -20,51 +20,56 @@ import (
 	ent "github.com/weaviate/weaviate/entities/vectorindex/hnsw"
 )
 
-func (h *hnsw) Compress(cfg ent.PQConfig) error {
-	h.shardedNodeLocks.RLock(0)
-	node := h.nodes[0]
-	h.shardedNodeLocks.RUnlock(0)
+func (h *hnsw) Compress(cfg ent.UserConfig) error {
+	h.compressActionLock.Lock()
+	defer h.compressActionLock.Unlock()
+	data := h.cache.All()
+	if cfg.PQ.Enabled {
+		h.shardedNodeLocks.RLock(0)
+		node := h.nodes[0]
+		h.shardedNodeLocks.RUnlock(0)
 
-	if node == nil {
-		return errors.New("data must be inserted before compress command can be executed")
-	}
+		if node == nil {
+			return errors.New("data must be inserted before compress command can be executed")
+		}
 
-	vec, err := h.vectorForID(context.Background(), node.id)
-	if err != nil {
-		return fmt.Errorf("infer vector dimensions: %w", err)
-	}
-	dims := len(vec)
+		vec, err := h.vectorForID(context.Background(), node.id)
+		if err != nil {
+			return fmt.Errorf("infer vector dimensions: %w", err)
+		}
+		dims := len(vec)
 
-	// segments == 0 (default value) means use as many segments as dimensions
-	if cfg.Segments <= 0 {
-		cfg.Segments = dims
-	}
+		// segments == 0 (default value) means use as many segments as dimensions
+		if cfg.PQ.Segments <= 0 {
+			cfg.PQ.Segments = dims
+		}
 
-	data := h.cache.All()
-	cleanData := make([][]float32, 0, len(data))
-	for _, point := range data {
-		if point == nil {
-			continue
+		cleanData := make([][]float32, 0, len(data))
+		for _, point := range data {
+			if point == nil {
+				continue
+			}
+			cleanData = append(cleanData, point)
 		}
```

Comment:
As I understand index of vector in cache is vector's `id`. Here all empty element are removed, but also indexes change. Is this index change safe? Are we sure that indexes are not lost or mixed later on?
(I see that compressor's cache is grown to size of cleaned up slice, but that may be to small for highest ids if some empty elements were indeed removed)

Reply:
`cleanData` is only used for fitting `KMeans`. Latter, the points are added using `data` instead of `cleanData` and this is when they are added to the cache. The two slices are needed since `KMeans` does expect a nil free array so the code is safe since the `id` is only needed for the cache where we do use `data` and not `cleanData`.